File: modules/21-llmops-eval/ts/03-regression-gate.ts
/**
 * Task 3 — Regression gate in CI  🟡
 *
 * What this teaches:
 *   - A CI gate fails the build (process.exit(1)) when a key metric drops
 *     below a threshold, preventing regressions from reaching production.
 *   - The same script runs in GitHub Actions and as a pre-push hook.
 *
 * How to run:
 *   pnpm tsx modules/21-llmops-eval/ts/03-regression-gate.ts \
 *       --results modules/21-llmops-eval/results/<run>.json \
 *       --metric avg_score --threshold 0.6
 *
 *   pnpm tsx modules/21-llmops-eval/ts/03-regression-gate.ts \
 *       --run-fresh --metric faithfulness --threshold 0.7
 *
 * Exit codes: 0 = pass, 1 = fail (metric below threshold).
 */

import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { getProvider } from "@learn-ai/llm-core";

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = resolve(__dirname, "../../../");
const DATA_DIR = resolve(REPO_ROOT, "modules/21-llmops-eval/data");
const EVAL_SET_PATH = resolve(DATA_DIR, "eval_set_v1.json");

const DEFAULT_THRESHOLDS: Record<string, number> = {
  avg_score: 0.60,
  pass_rate: 0.60,
  faithfulness: 0.70,
};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface ResultsSummary {
  avg_score?: number;
  pass_rate?: number;
  avg_latency_ms?: number;
  faithfulness?: number;
}

interface GraderResult {
  grader: string;
  score: number;
  passed: boolean;
}

interface CaseData {
  grader_results?: GraderResult[];
  [key: string]: unknown;
}

interface ResultsFile {
  summary?: ResultsSummary;
  cases?: CaseData[];
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Step 1 — Load results file
// ---------------------------------------------------------------------------

/**
 * Parse a JSON results file produced by task 1 or 2.
 *
 * TODO 1: Read the file as UTF-8 text and parse the JSON.
 * Throw a descriptive error if the file doesn't exist.
 * Return the parsed object as ResultsFile.
 */
function loadResults(path: string): ResultsFile {
  // TODO: implement loadResults
  throw new Error("TODO: implement loadResults");
}

// ---------------------------------------------------------------------------
// Step 2 — Extract metric
// ---------------------------------------------------------------------------

/**
 * Pull the numeric value of `metric` from the results dict.
 *
 * TODO 2a: Try results.summary?.[metric].
 * TODO 2b: For 'faithfulness', average the llm_judge grader scores
 *          across results.cases.
 * TODO 2c: Throw if the metric can't be found.
 * Return the number.
 */
function extractMetric(results: ResultsFile, metric: string): number {
  // TODO: implement extractMetric
  throw new Error("TODO: implement extractMetric");
}

// ---------------------------------------------------------------------------
// Step 3 — Run fresh eval
// ---------------------------------------------------------------------------

/**
 * Run a quick eval and return a ResultsFile-compatible object.
 *
 * TODO 3a: Load eval set from EVAL_SET_PATH.
 * TODO 3b: For each case, call provider.chat() with context-stuffed prompt.
 * TODO 3c: Score faithfulness via LLM-judge (0–10 → /10).
 * TODO 3d: Return { summary: { avg_score, pass_rate, faithfulness }, cases }.
 */
async function runFreshEval(
  provider: ReturnType<typeof getProvider>,
): Promise<ResultsFile> {
  // TODO: implement runFreshEval
  throw new Error("TODO: implement runFreshEval");
}

// ---------------------------------------------------------------------------
// Step 4 — Gate check
// ---------------------------------------------------------------------------

/**
 * Log pass/fail and return true if value >= threshold.
 *
 * TODO 4: Print one clear line tagged PASS or FAIL that names the metric, its
 *         value, the comparison, and the threshold.
 * Return boolean.
 */
function checkGate(value: number, threshold: number, metric: string): boolean {
  // TODO: implement checkGate
  throw new Error("TODO: implement checkGate");
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

function parseArgs(): {
  resultsPath: string | null;
  runFresh: boolean;
  metric: string;
  threshold: number | null;
} {
  const argv = process.argv.slice(2);
  let resultsPath: string | null = null;
  let runFresh = false;
  let metric = "avg_score";
  let threshold: number | null = null;

  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--results" && argv[i + 1]) resultsPath = argv[++i];
    if (argv[i] === "--run-fresh") runFresh = true;
    if (argv[i] === "--metric" && argv[i + 1]) metric = argv[++i];
    if (argv[i] === "--threshold" && argv[i + 1]) threshold = parseFloat(argv[++i]);
  }
  return { resultsPath, runFresh, metric, threshold };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const { resultsPath, runFresh, metric, threshold: thresholdArg } = parseArgs();
  const threshold = thresholdArg ?? DEFAULT_THRESHOLDS[metric] ?? 0.6;

  let results: ResultsFile;

  if (runFresh) {
    const provider = getProvider();
    console.log(`Running fresh eval with ${provider.name}/${provider.chatModel}...`);
    results = await runFreshEval(provider);
  } else if (resultsPath) {
    results = loadResults(resultsPath);
  } else {
    console.error("Provide --results <path> or --run-fresh.");
    process.exit(1);
  }

  const value = extractMetric(results, metric);
  const passed = checkGate(value, threshold, metric);
  process.exit(passed ? 0 : 1);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: modules/21-llmops-eval/py/03_regression_gate.py
"""
Task 3 — Regression gate in CI  🟡

What this teaches:
  - A CI gate fails the build (non-zero exit) when a key metric drops below a
    threshold, preventing regressions from reaching production.
  - The same script works both as a standalone CLI check AND as a pre-push hook
    (see .github/workflows/eval-gate.yml and the husky note in the README).
  - Thresholds belong in config, not hard-coded — so different teams/models
    can set different bars.

How to run:
  # Run against a pre-built results file:
  uv run python modules/21-llmops-eval/py/03_regression_gate.py \\
      --results modules/21-llmops-eval/results/<run>.json \\
      --metric avg_score --threshold 0.6

  # Run a fresh eval then gate:
  uv run python modules/21-llmops-eval/py/03_regression_gate.py \\
      --run-fresh --metric faithfulness --threshold 0.7

  # Non-zero exit if threshold is not met (for CI):
  # The script exits with code 1 on failure, 0 on pass.
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any

from llm_core import ChatMessage, ChatOptions, get_provider

REPO_ROOT = Path(__file__).resolve().parents[3]
DATA_DIR = REPO_ROOT / "modules" / "21-llmops-eval" / "data"
RESULTS_DIR = REPO_ROOT / "modules" / "21-llmops-eval" / "results"
EVAL_SET_PATH = DATA_DIR / "eval_set_v1.json"

# Default thresholds — override per metric via --threshold
DEFAULT_THRESHOLDS: dict[str, float] = {
    "avg_score": 0.60,
    "pass_rate": 0.60,
    "faithfulness": 0.70,   # from LLM-judge; requires --run-fresh
}


# ---------------------------------------------------------------------------
# Step 1 — Load a results file (from task 1 or 2)
# ---------------------------------------------------------------------------

def load_results(path: Path) -> dict[str, Any]:
    """Parse a JSON results file produced by task 1 or 2.

    TODO 1: Open path, parse JSON, return the dict.
    Raise FileNotFoundError with a helpful message if the path doesn't exist.
    """
    raise NotImplementedError("TODO: implement load_results")


# ---------------------------------------------------------------------------
# Step 2 — Extract the metric value
# ---------------------------------------------------------------------------

def extract_metric(results: dict[str, Any], metric: str) -> float:
    """Pull the numeric value of `metric` from the results dict.

    The results structure has a 'summary' key with: avg_score, pass_rate,
    avg_latency_ms. It may also have per-case grader scores.

    TODO 2a: Try results['summary'][metric].
    TODO 2b: If not there, check for a per-case 'faithfulness' average:
             iterate results['cases'], find grader_results with grader=='llm_judge',
             average their scores, and use that.
    TODO 2c: Raise ValueError if the metric cannot be found.
    Return the float value.
    """
    raise NotImplementedError("TODO: implement extract_metric")


# ---------------------------------------------------------------------------
# Step 3 — Run a fresh eval (optional, reuses task 1 logic)
# ---------------------------------------------------------------------------

def run_fresh_eval(provider: Any) -> dict[str, Any]:
    """Run a quick eval and return a results-dict compatible with extract_metric.

    This is a stripped version of task 1's runner (no file write needed —
    the gate only cares about the numbers).

    TODO 3a: Load eval set from EVAL_SET_PATH.
    TODO 3b: For each case, call provider.chat() with a system prompt that
             includes the context.
    TODO 3c: Score faithfulness with an LLM-judge call (score 0–10, /10).
    TODO 3d: Return {'summary': {'avg_score': ..., 'pass_rate': ...,
                                  'faithfulness': ...},
                      'cases': [...]}.
    Tip: faithfulness = avg of LLM-judge scores across all cases.
    """
    raise NotImplementedError("TODO: implement run_fresh_eval")


# ---------------------------------------------------------------------------
# Step 4 — Gate check
# ---------------------------------------------------------------------------

def check_gate(metric_value: float, threshold: float, metric: str) -> bool:
    """Return True if metric_value >= threshold, else print failure and return False.

    TODO 4: Print one clear line tagged PASS or FAIL that names the metric, its
            value, the comparison, and the threshold, so CI logs are scannable.
    Return True on pass, False on fail.
    """
    raise NotImplementedError("TODO: implement check_gate")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        description="Regression gate — exits non-zero if a metric drops below threshold."
    )
    parser.add_argument(
        "--results", type=Path, default=None,
        help="Path to a results JSON from task 1/2. Mutually exclusive with --run-fresh.",
    )
    parser.add_argument(
        "--run-fresh", action="store_true",
        help="Run a fresh eval before checking the gate.",
    )
    parser.add_argument(
        "--metric", default="avg_score",
        choices=list(DEFAULT_THRESHOLDS) + ["faithfulness"],
        help="Which metric to gate on.",
    )
    parser.add_argument(
        "--threshold", type=float, default=None,
        help="Override the default threshold for this metric.",
    )
    args = parser.parse_args()

    threshold = args.threshold if args.threshold is not None else DEFAULT_THRESHOLDS.get(args.metric, 0.6)

    if args.run_fresh:
        provider = get_provider()
        print(f"Running fresh eval with {provider.name}/{provider.chat_model}...")
        results = run_fresh_eval(provider)
    elif args.results:
        results = load_results(args.results)
    else:
        parser.error("Provide --results <path> or --run-fresh.")

    value = extract_metric(results, args.metric)
    passed = check_gate(value, threshold, args.metric)

    sys.exit(0 if passed else 1)


if __name__ == "__main__":
    main()
